nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarProducerProcessor.java

```diff
@@ -0,0 +1,472 @@
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+
+    public static final PropertyDescriptor ASYNC_ENABLED = new PropertyDescriptor.Builder()
+            .name("ASYNC_ENABLED")
+            .displayName("Async Enabled")
+            .description("Control whether the messages will be sent asyncronously or not. Messages sent"
+                    + " syncronously will be acknowledged immediately before processing the next message, while"
+                    + " asyncronous messages will be acknowledged after the Pulsar broker responds. Running the"
+                    + " processor with async enabled will result in increased the throughput at the risk of potential"
+                    + " duplicate data being sent to the Pulsar broker.")
+            .required(true)
+            .allowableValues("true", "false")
+            .defaultValue("false")
+            .build();
+
+    public static final PropertyDescriptor MAX_ASYNC_REQUESTS = new PropertyDescriptor.Builder()
+            .name("MAX_ASYNC_REQUESTS")
+            .displayName("Maximum Async Requests")
+            .description("The maximum number of outstanding asynchronous publish requests for this processor. "
+                    + "Each asynchronous call requires memory, so avoid setting this value to high.")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("50")
+            .build();
+
+    public static final PropertyDescriptor BATCHING_ENABLED = new PropertyDescriptor.Builder()
+            .name("BATCHING_ENABLED")
+            .displayName("Batching Enabled")
+            .description("Control whether automatic batching of messages is enabled for the producer. "
+                    + "default: false [No batching] When batching is enabled, multiple calls to "
+                    + "Producer.sendAsync can result in a single batch to be sent to the broker, leading "
+                    + "to better throughput, especially when publishing small messages. If compression is "
+                    + "enabled, messages will be compressed at the batch level, leading to a much better "
+                    + "compression ratio for similar headers or contents. When enabled default batch delay "
+                    + "is set to 10 ms and default batch size is 1000 messages")
+            .required(true)
+            .allowableValues("true", "false")
+            .defaultValue("false")
+            .build();
+
+    public static final PropertyDescriptor BATCHING_MAX_MESSAGES = new PropertyDescriptor.Builder()
+            .name("BATCHING_MAX_MESSAGES")
+            .displayName("Batching Max Messages")
+            .description("Set the maximum number of messages permitted in a batch. default: "
+                    + "1000 If set to a value greater than 1, messages will be queued until this "
```

Comment:
i'd recommend not doing any queuing until this value is reached.  Rather use this as a hard limit for how many go in a batch but the min that goes should be the max this is available/waiting/ready at a given time.  Otherwise you're introducing false latency

Reply:
All of the "batching" specified by these properties actually occurs within the Pulsar client itself, and are not kept in NiFi anywhere.  This property merely exposes that setting to the end user if they wish to use that feature to achieve better write throughput to Pulsar.

http://pulsar.apache.org/api/client/org/apache/pulsar/client/api/ProducerBuilder.html#enableBatching-boolean-